blockstore/splitstore/markset_badger.go

```diff
@@ -0,0 +1,225 @@
+package splitstore
+
+import (
+	"os"
+	"path/filepath"
+	"sync"
+
+	"golang.org/x/xerrors"
+
+	"github.com/dgraph-io/badger/v2"
+	"github.com/dgraph-io/badger/v2/options"
+	"go.uber.org/zap"
+
+	cid "github.com/ipfs/go-cid"
+)
+
+type BadgerMarkSetEnv struct {
+	path string
+}
+
+var _ MarkSetEnv = (*BadgerMarkSetEnv)(nil)
+
+type BadgerMarkSet struct {
+	mx      sync.RWMutex
+	cond    sync.Cond
+	pend    map[string]struct{}
+	writing map[int]map[string]struct{}
+	writers int
+	seqno   int
+
+	db   *badger.DB
+	path string
+}
+
+var _ MarkSet = (*BadgerMarkSet)(nil)
+
+var badgerMarkSetBatchSize = 16384
+
+func NewBadgerMarkSetEnv(path string) (MarkSetEnv, error) {
+	msPath := filepath.Join(path, "markset.badger")
+	err := os.MkdirAll(msPath, 0755) //nolint:gosec
+	if err != nil {
+		return nil, xerrors.Errorf("error creating markset directory: %w", err)
+	}
+
+	return &BadgerMarkSetEnv{path: msPath}, nil
+}
+
+func (e *BadgerMarkSetEnv) Create(name string, sizeHint int64) (MarkSet, error) {
+	path := filepath.Join(e.path, name)
```

Comment:
Do we need to add a salt to this filename on the caller side? `Create` always get called with a fixed name `live`. This means that:

1. on restart, we will open an _existing_ markset. However, this is unsafe because `SyncWrites` below is false!
2. a failure to `Close` and `Remove` will end up opening a previously existing database.

If we always want to operate on the same database, we need to ensure that reopening it will always be safe. However, there's probably little value in doing that because we do not resume marking on a process restart. So the recommendation here is to add a salt.

OTOH, an interruption could lead to leftover/abandoned markset databases.

Reply:
No, on restart we just delete everything -- marksets are transient.

Reply:
Correction: when we reopen the markset, it's deleted.

So basically there is no safety concerns from close errors. The close implementation deletes the thing, and when we reopen it we delete again just to be sure we didn't have left overs.

Basically the on-disk markset is completely transient, we don't care at all about persisting it once we are done using it.

Reply:
On the other hand, using a salt would leave us with leftover marksets we can't consistently delete on restart, basically garbage on disk.
So using a salt is not the right thing to do here, and the code behaves as desired afaict.

Reply:
Ack. Makes sense.